Lib/test/test_dataclasses.py

```diff
@@ -2902,6 +2902,59 @@ class A:
         self.assertEqual(obj.a, 'a')
         self.assertEqual(obj.b, 'b')
 
+    def test_frozen_slots_raises_correct_error(self):
+        # bpo-45897
+        @dataclass(slots=True, frozen=True)
+        class Point:
+            x: int
+            y: int
+
+        p = Point(1, 2)
+        self.assertEqual(p.x, 1)
+        self.assertEqual(p.y, 2)
+        with self.assertRaises(FrozenInstanceError):
+            p.x = 2
+
+        with self.assertRaises(AttributeError) as ex:
+            p.z = 5
+        type_of_caught_exception = type(ex.exception)
+
+        # Verify that the caught exception is `AttributeError` exactly,
+        # and *not* a subclass of `AttributeError` such as `FrozenInstanceError`.
+        #
+        # We need to ensure that this error is being caused by the standard machinery that
+        # prevents new attributes being added that aren't specified in __slots__.
+        #
+        # We *do not* want the error to be caused by anything related to dataclasses.
+        self.assertIs(
+            type_of_caught_exception, 
+            AttributeError, 
+            msg=f"expected 'AttributeError', got {type_of_caught_exception.__name__!r}"
+        )
+
+        @dataclass(frozen=True)
+        class DataclassSubclassOfPointWithNoSlots(Point): pass
+
+        d = DataclassSubclassOfPointWithNoSlots(1, 2)
+        with self.assertRaises(FrozenInstanceError):
+            d.x = 2
+
+        # The subclass is frozen, so any attempt to set an attribute should raise an error.
+        # The subclass does not define slots, however,
+        # so the error should be `FrozenInstanceError` here rather than `AttributeError`.
```

Comment:
Wrap-at-79-chars-nitpick:

```suggestion
        # The subclass is frozen, so any attempt to set an attribute should
        # raise an error. The subclass does not define slots, however, so the
        # error should be `FrozenInstanceError` here rather than
        # `AttributeError`.
```